### ISSUE TITLE
Distributed Training of the Decoder

### ISSUE THREAD

**Issue by Veldrovive** [lucidrains/DALLE2-pytorch#114]
With the single GPU training script in place, I think it is time to work on a multi-gpu version. There has been discussion in #103 of strategies for backend agnostic training so we will have to watch that to keep parity.

Preliminary testing on this branch [branch](https://github.com/Veldrovive/DALLE2-pytorch/tree/unet_ddp) seems to be working. The strategy used there is to have the decoder wrap the its unets with ddp since wrapping the decoder itself doesn't work because the forward call in that case is done using a non-ddp unet and the gradient hook does not seem to fire. With wrapping the unets themselves the weights stay in alignment so it appears to work.

No larger scale tests have been conducted, but I wanted to open this before going further so that if there is feedback on the strategy or something that needs to change, we can tackle that.

**Comment by nousr** [comment]
Awesome! Have you tried out huggingface accelerate? 

I've been testing it out locally and it seems pretty handy. The biggest benefit imo is that its fairly easy to write code that will work for both single gpu and distributed training (letting people who don't have access to multi-gpu pods to still contribute, in theory). 

In an effort to still remain a little more agnostic, I've been wrapping a lot of the `accelerator` commands with the prior--just so that it **can** be used without accelerator; I've been messing around with it in this [branch](https://github.com/nousr/DALLE2-pytorch/blob/f42a95a6748e67ade9c1b88ffc5a642cf121d498/dalle2_pytorch/trainer.py#L249) if you'd like to take a look.

This is more or less the biggest change you need to make is this, and the rest is relatively straightforward as the API is pretty intuitive.

```python
def distribute(self, accelerator):


    # ensure you can actually distribute
    assert isinstance(accelerator, Accelerator)


    self.print("Preparing HuggingFace Accelerator...")


    # give this trainer an accelerator
    self.accelerator = accelerator


    # distribute the work
    self.diffusion_prior, self.optimizer = self.accelerator.prepare(self.diffusion_prior, self.optimizer)
```

**Comment by Veldrovive** [comment]
Rom also suggested trying out accelerate. In a small test, it appears to work when wrapping the Decoder itself which is strange because it is just using DDP on the backend so either it is doing something special or I missed something in my own test. I'll have to see later though cause I should probably do my real job.

**Comment by Veldrovive** [comment]
Using accelerate, I created a version of the decoder trainer that is compatible with both single and multi process training. There are still somethings to handle (using accelerate for amp and handling grad norm clipping), but it is functional. Test branch [here](https://github.com/Veldrovive/DALLE2-pytorch/tree/accelerate_trainer).
